go/vt/vttablet/tabletmanager/init_tablet.go

```diff
@@ -122,22 +125,17 @@ func (agent *ActionAgent) InitTablet(port, gRPCPort int32) error {
 		}
 	}
 
-	// See if we need to add the tablet's cell to the shard's cell list.
-	if !si.HasCell(agent.TabletAlias.Cell) {
-		if err := agent.withRetry(ctx, "updating Cells list in Shard if necessary", func() error {
-			si, err = agent.TopoServer.UpdateShardFields(ctx, *initKeyspace, shard, func(si *topo.ShardInfo) error {
-				if si.HasCell(agent.TabletAlias.Cell) {
-					// Someone else already did it.
-					return topo.NewError(topo.NoUpdateNeeded, agent.TabletAlias.String())
-				}
-				si.Cells = append(si.Cells, agent.TabletAlias.Cell)
-				return nil
-			})
-			return err
-		}); err != nil {
-			return vterrors.Wrap(err, "couldn't add tablet's cell to shard record")
-		}
+	// Rebuild keyspace graph if this the first tablet in this keyspace/cell
+	_, err = agent.TopoServer.GetSrvKeyspace(ctx, agent.TabletAlias.Cell, *initKeyspace)
+	switch {
+	case err == nil:
+		// NOOP
+	case topo.IsErrType(err, topo.NoNode):
+		err = topotools.RebuildKeyspaceLocked(ctx, logutil.NewConsoleLogger(), agent.TopoServer, *initKeyspace, []string{agent.TabletAlias.Cell})
```

Comment:
I'm not seeing a lock on the keyspace around this call to RebuildKeyspaceLocked()?

If there isn't one, seems like there could also be risk that the topo changed between the `GetSrvKeyspace` above and here.

Reply:
Good catch. Addressed in: Addressed in: https://github.com/vitessio/vitess/pull/4760